File: LeetCode/150 Ques/arrays/three_sum.go
/*
Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]]
such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

Notice that the solution set must not contain duplicate triplets.



Example 1:

Input: nums = [-1,0,1,2,-1,-4]
Output: [[-1,-1,2],[-1,0,1]]


Example 2:

Input: nums = []
Output: []


Example 3:

Input: nums = [0]
Output: []
*/

package main

import (
	"fmt"
)

func three_sum(arr []int, target int) [][]int {
	n := len(arr)
	res := [][]int{}

	for i := 0; i < n; i++ {
		k := target - arr[i]
		mp := make(map[int]bool)
		for j := i + 1; j < n; j++ {
			if _, ok := mp[k-arr[j]]; ok {
				res = append(res, []int{arr[i], arr[j], k - arr[j]})
			} else {
				mp[arr[j]] = true
			}
		}
	}
	temp := [][]int{}
	for i := 0; i < len(res); i++ {
		for j := i + 1; j < len(res); j++ {
			if (res[i][0] == res[j][0] || res[i][1] == res[j][0] || res[i][2] == res[j][0]) &&
				(res[i][0] == res[j][1] || res[i][1] == res[j][1] || res[i][2] == res[j][1]) &&
				(res[i][0] == res[j][2] || res[i][1] == res[j][2] || res[i][2] == res[j][2]) {
				res[j][0] = -99999
				res[j][1] = -99999
				res[j][2] = -99999
			}
		}
	}

	for i := 0; i < len(res); i++ {
		if res[i][0] != -99999 {
			temp = append(temp, res[i])
		}
	}

	return temp
}

func main() {
	arr := []int{-1, 0, 1, 2, -1, -4}
	fmt.Println(three_sum(arr, 0))
}

File: LeetCode/neetcode 150/two pointers/three_sum.cc
class Solution {
public:
    vector<vector<int>> res;
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        for (int i = 0; i < nums.size(); i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            two_sum(nums, nums[i], i + 1);
        }
        return res;
    }

    void two_sum(vector<int>& nums, int curr, int i) {
        int l = i;
        int r = nums.size() - 1;
        vector<int> temp;
        while (l < r) {
            int sum = nums[l] + nums[r] + curr;
            if (sum == 0) {
                res.push_back({curr, nums[l], nums[r]});
                l++;
                r--;
                while (l < r && nums[l] == nums[l - 1]) l++;
            } else if (sum > 0) {
                r--;
            } else {
                l++;
            }
        }
    }
};
